http_server.py

```diff
@@ -0,0 +1,107 @@
+import socket
+import os
+import sys
+
+
+ROOT = "./webroot/"
+ADDR = ("127.0.0.1", 8000)
+server = socket.socket(
+    socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_IP
+)
+server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+
+def resolve_uri(uri):
+    if "../" in uri:
+        raise UserWarning("403 Forbidden")
+
+    if os.path.isfile(uri):
+        TYPE = uri.split(".")[-1].lower()
+        if TYPE in ('PNG', 'JPG', 'JPEG', 'GIF'):
+            TYPE = 'image/' + TYPE
+        else:
+            TYPE = 'text/' + TYPE
+        BODY = open(uri).read()
+
+    elif os.path.isdir(uri):
+        TYPE = "text/html"
+        BODY = "<!DOCTYPE HTML><html><head><title>Directory</title></head><h1>Directory</h1>"
+        for root, dirs, files in os.walk(uri):
+            BODY += "<h3>" + root + "</h3>"
+            BODY += "<ul>"
+            for f in files:
+                BODY += "<li>" + f + "</li>"
+            BODY += "</ul>"
+        BODY += "</body></html>"
+
+    else:
+        raise IOError("404 Not Found")
+    return TYPE, BODY
+
+
+def parse_request(request):
+    try:
+        request = request.split("\r\n\r\n", 1)
+        req = request[0].split("\r\n")
+        for i, r in enumerate(req):
+            req[i] = r.split()
+        method = req[0][0].upper()
+        uri = req[0][1]
+        proto = req[0][2].upper()
+        headers = {}
+        for line in req[1:]:
+            headers[line[0].upper()] = line[1:]
+        if method == "GET":
+            if proto == "HTTP/1.1" and "HOST:" in headers:
+                return ROOT + uri
```

Comment:
because of this, your `parse_request` method is now aware of an implementation detail of the server.  I'm sure that doesn't hurt anything now, but it muddies the lines of control in a way that eventually will lead to trouble.  I'd pass back the URI as requested rather than altering it in this function and combine it with the root later, perhaps in `resolve_uri`